docs/device-assignment.md

```diff
@@ -0,0 +1,130 @@
+# Virtual Machine Device Assignment
+
+## Scope
+
+For the first version of the document, the scope is reduced PCI device assignment.
+
+## Introduction
+
+Device assignment is a feature that allows virtual machines (VMs) to take advantage of hardware capabilities and performance available on the hypervisor (host/node). Unlike traditional VM resources such as virtual CPUs (vCPUs) or memory, assignment-capable hardware devices are limited resource that cannot be overcommitted.
+
+### Device Assignment in Kubernetes
+
+The first version of device assignment was added to Kubernetes in version 1.3, and worked specifically with NVIDIA GPUs. Starting in version 1.8, the feature is slowly being deprecated by system called device plugins (DPIs). Both versions of assignment share similar traits: the node exposes a scalar resource indicating quantity of the device, and pods request the resource either directly or through node affinity rules. The legacy NVIDIA plugin was only able to expose NVIDIA GPU. In case of DPI, the devices are reported and managed through separate daemon(s), where exposed devices depend on the plugin's code. Both versions of the assignment worked with assumption that the container requires several node paths and container paths. The paths contained endpoints needed for actual GPU compute.
+
+The system is required to provide the devices in working state - having the proper driver installed and bound to the device.
+
+### (PCI) Device Assignment in KVM/QEMU (Linux)
+
+Device assignment in KVM/QEMU has different requirements than in case of containers that stem from the higher level of isolation provided by VMs when compared to containers.
+
+When using VFIO (Virtual Function I/O; modern device assignment framework), the system must be IOMMU (I/O Memory Management Unit) capable. That roughly translates to having Intel VT-d or AMD-Vi virtualization extensions, the kernel booted with IOMMU module, chipset that happens to work, ACS (PCI Access Control Services) capable CPU (= business class) for proper level of isolation, and possibly more.
+
+The concept of IOMMU deserves an explanation on it's own: from high level point of view, the devices in the system may be isolated from each other. If the isolation is provided by the IOMMU, the device or devices within that isolated area are treated as one group. Imagine a system without any isolation - in that case, every device would be in one IOMMU group. On the other hand, extreme level of isolation between devices yields one IOMMU group for each device. In practice, the number of devices in a group varies quite a lot. Consumer level GPUs usually contain at least an audio device (that handles HDMI audio), so that makes a group of 2 devices. Network interface cards' ports are usually grouped into groups of 1-4 devices.
```

Comment:
@mpolednik In the context of your problem description at https://github.com/mpolednik/linux-vfio-k8s-dpi,  do you plan to let your plugin assign the whole IOMMU group for the start? Just allocating the whole IOMMU group does not sound so bad to me. Especially since the plugin would report that it assigned all devices of that IOMMU to one pod, which means that the resources count on the node object would adjust.

The worst thing I see would be, that in some edge cases one VM will be rejected/evicted from the node it got scheduled to, and needs to be rescheduled. Are there any other disadvantages?

Reply:
@rmohr For a start, I've found assigning the whole group to be *really* hard solution. The communication between Kubelet and DPI is pretty limited in 1.8: Kubelet requests a device, DPI delivers. If DPI returns more devices than Kubelet originally requested, their host/container paths are mounted but the Pod resource is not updated with extra devices (because Kubelet doesn't have enough info to fill that in for us).

There's one more mechanism in the DPI framework, and that is device healthiness. The plugin could mark devices that were bound to vfio-pci but not requested by the pod as unhealthy, allowing kubelet to update the Node capacity accordingly. There's quite a caveat though: Kubelet doesn't communicate when device is "released", so the DPI can't mark the device healthy again 😞 .

In the far future, I'd like to solve this on Kubelet level as long as it doesn't break the container device vision. For short to mid term, I'm exploring 2 solutions: the device plugin detecting when device is not used (will require some driver magic I guess, if possible) or finding a way to inject the extra devices into the Pod spec. What I want to avoid is shifting the burden of managing IOMMU groups to the cloud admins.